src/core/smoothing/horizon.rs

```diff
@@ -5,89 +5,80 @@ use super::*;
+    let z_axis = nalgebra::Vector3::<f64>::z_axis();
+
+    let corrected_transform = q.to_rotation_matrix() * Rotation3::from_axis_angle(&y_axis, -std::f64::consts::FRAC_PI_2) * Rotation3::from_axis_angle(&z_axis, -std::f64::consts::FRAC_PI_2);
+    // since this coincides with roll axis, the roll is neglected when transformed back
+    let axis_transformed = corrected_transform * axis;
+
+    let pitch = (axis_transformed.z).asin();
+    let yaw = axis_transformed.y.simd_atan2(axis_transformed.x) - std::f64::consts::FRAC_PI_2;
+    
+    from_euler_yxz(pitch, roll_correction, yaw)
+}
+
 #[derive(Clone)]
 pub struct HorizonLock {
-    pub time_constant: f64,
+    pub lock_enabled: bool,
     pub horizonlockpercent: f64,
     pub horizonroll: f64,
 }
 
+
 impl Default for HorizonLock {
     fn default() -> Self { Self {
-        time_constant: 0.25,
-        horizonlockpercent: 0.0,
+        lock_enabled: true,
+        horizonlockpercent: 100.0,
         horizonroll: 0.0,
     } }
 }
 
-impl SmoothingAlgorithm for HorizonLock {
-    fn get_name(&self) -> String { "Lock horizon".to_owned() }
-
-    fn set_parameter(&mut self, name: &str, val: f64) {
-        match name {
-            "time_constant" => self.time_constant = val,
-            "horizonroll" => self.horizonroll = val,
-            "horizonlockpercent" => self.horizonlockpercent = val,
-            _ => log::error!("Invalid parameter name: {}", name)
-        }
+impl HorizonLock {
+    pub fn set_horizon(&mut self, lock_percent: f64, roll: f64) {
+        self.horizonroll = roll;
+        self.horizonlockpercent = lock_percent;
+        self.lock_enabled = self.horizonlockpercent > 1e-6;
     }
-    fn get_parameters_json(&self) -> serde_json::Value {
-        serde_json::json!([
-            {
-                "name": "time_constant",
-                "description": "Smoothness",
-                "type": "SliderWithField",
-                "from": 0.01,
-                "to": 10.0,
-                "value": self.time_constant,
-                "default": 0.25,
-                "unit": "s"
-            }
-        ])
-    }
-    fn get_status_json(&self) -> serde_json::Value {
-        serde_json::json!([])
-    }
-
-    fn get_checksum(&self) -> u64 {
+    pub fn get_checksum(&self) -> u64 {
         let mut hasher = std::collections::hash_map::DefaultHasher::new();
-        hasher.write_u64(self.time_constant.to_bits());
-        hasher.write_u64(self.horizonroll.to_bits());
         hasher.write_u64(self.horizonlockpercent.to_bits());
+        hasher.write_u64(self.horizonroll.to_bits());
         hasher.finish()
     }
 
-    fn smooth(&mut self, quats: &TimeQuat, duration: f64, _stabilization_params: &StabilizationParams) -> TimeQuat { // TODO Result<>?
-        if quats.is_empty() || duration <= 0.0 { return quats.clone(); }
-
-        let sample_rate: f64 = quats.len() as f64 / (duration / 1000.0);
+    pub fn lockquat(&mut self, q: UnitQuaternion<f64>) -> UnitQuaternion<f64> {
```

Comment:
I would just name it `lock`

Reply:
ah it's for a single quat? then `lock_single`

Reply:
`lockquat` is just for one quat while `lock` below handles TimeQuat